Oriented bounding box & frustum culling optimizations

kainino0x opened the pull request:
(Do not merge. Will be tested with 3D buildings first.)

In review & editing.

* [x] Further testing of rectangles which span the equator
* [x] Overall (rendering and network) performance benchmarking.  Update the TODO in `CHANGES.md`.

Future:

* [x] Hook up to heightmapped terrain (and flat terrain, if that's separate); not just quantized meshes
* [ ] Hook up to 2D and Columbus view rendering modes

pjcozzi commented on Source/Core/Matrix3.js:
```diff
@@ -1355,6 +1355,16 @@ define([
+     * @type {Matrix3}
+     * @constant
+     */
+    Matrix3.ZERO = freezeObject(new Matrix3(0.0, 0.0, 0.0,
```
For consistency, add this to `Matrix2` and `Matrix4`.  Mention these in CHANGES.md.

kainino0x replied:
Where in CHANGES.md should I put it? Since I'm assuming the 1.10 section is final.

pjcozzi replied:
Make a new section for 1.11.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,330 @@
+        }
+
+        if (!defined(positions) || positions.length === 0) {
+            result.center = Cartesian3.ZERO;
```
Use `Cartesian3.clone` and `Matrix3.clone`.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,330 @@
+     *                      on the opposite side, and {@link Intersect.INTERSECTING} if the box
+     *                      intersects the plane.
+     */
+    OrientedBoundingBox.planeIntersect = function(box, plane) {
```
Probably rename this to `intersectPlane`.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,330 @@
+     * @inheritDoc OrientedBoundingBox.planeIntersect
+     * @deprecated
+     */
+    OrientedBoundingBox.prototype.intersect = function(plane) {
```
Don't include this.  Instead rename the other types with an `intersect` function to `intersectPlane` and then deprecate their `intersect` function. See https://github.com/AnalyticalGraphicsInc/cesium/wiki/Deprecation-Guide

pjcozzi replied:
Also, looks like we can deprecate the old OOBB, right?

kainino0x replied:
Should I do the `intersect` -> `intersectPlane` rename in a separate PR and rebase onto that?

I'll deprecate the old OOBB in this PR.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,330 @@
+    };
+
+    /**
+     * @inheritDoc OrientedBoundingBox.planeIntersect
```
We have never used `@inheritDoc` AFAIK, but does this work in this case where the arguments are different since `this` is implicit here?

kainino0x replied:
I don't think it does. I meant to ask about this, though. Should I just copy-paste the other documentation?

pjcozzi replied:
Kind of.  Look at the [Matrix4](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Matrix4.js) `clone` functions.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,330 @@
+
+        var center = box.center;
+        // plane is used as if it is its normal; the first three components are assumed to be normalized
+        var radEffective = Math.abs(Cartesian3.dot(plane, Matrix3.getColumn(box.halfAxes, 0, scratchCartesian1)))
```
Let's wait until we profile, but we could micro-optimize this a lot.

Also, make sure to profile for fps in a Cesium build with `DeveloperError` removed (`combineRelease` should do it I think), see https://github.com/AnalyticalGraphicsInc/cesium/wiki/Contributor%27s-Guide

pjcozzi commented on Specs/Core/OrientedBoundingBoxSpec.js:
```diff
@@ -0,0 +1,417 @@
+/*global defineSuite*/
+defineSuite([
```
Did you run code coverage?  What is it?

kainino0x replied:
It's at 96% now, but one is the function we're removing and the others will be easy to cover.

pjcozzi commented on Specs/Core/OrientedBoundingBoxSpec.js:
```diff
@@ -0,0 +1,417 @@
+        planeIntersectTestCornersEdgesFaces(Cartesian3.ZERO, Matrix3.fromScale(new Cartesian3(0.0, 0.0, 0.0), m));
+    });
+
+    //it('planeIntersect works with negatively scaled box', function() {
```
We generally don't keep commented out code.

kainino0x replied:
I didn't completely delete this because I wasn't sure if it was worth testing. However, I'm pretty sure it isn't.

pjcozzi commented on Specs/Core/OrientedBoundingBoxSpec.js:
```diff
@@ -0,0 +1,417 @@
+        planeIntersectTestCornersEdgesFaces(new Cartesian3(-5.1, 0.0, 0.1), n);
+    });
+
+    // TODO: test OrientedBoundingBox.planeIntersect more
```
We don't merge code with `TODOs`.  Is this still needed?

kainino0x replied:
No, I can remove that.

pjcozzi commented on Specs/Core/OrientedBoundingBoxSpec.js:
```diff
@@ -0,0 +1,417 @@
+
+        // Tests against faces
+
+        pl = planeNormXform(+1.0, +0.0, +0.0,          + 0.50001); if (pl) { expect(box.planeIntersect(pl)).toEqual(Intersect.INSIDE); }
```
There's extra whitespace in this and the following lines.

kainino0x replied:
I just had it like this since I was doing a lot of block editing across all of these lines. It should be removed now that that's done?

pjcozzi replied:
Please.

I would also drop the leading `+` and just use a space if you want it to line up vertically with the others.

pjcozzi commented:
This fails JSHint, see https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/65273151

pjcozzi commented:
Folks - this is the start of a pull request for more exact view frustum culling so we do not plan to merge this until we have it hooked up to terrain.

kainino0x commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,340 @@
+    };
+
+    /**
+     * XXX Temporary compatibility function - remove after rebasing ono the rename from intersect to intersectPlane.
```
TODO for self: remove when rebasing onto global rename from intersect to intersectPlane

pjcozzi commented:
> Should I do the intersect -> intersectPlane rename in a separate PR and rebase onto that?

It's fine to include it in this PR.

kainino0x commented:
TODO: Since we're changing `intersect` to `intersectPlane` anyway, consider replacing the `Cartesian4 plane` argument with an actual `Plane plane`.

pjcozzi commented:
+1 on changing to `Plane`.  This code was been written before `Plane`.

pjcozzi commented:
Just double check how wide of a breaking change it becomes.

kainino0x commented:
@pjcozzi `CullingVolume` exposes a {Cartesian4[]} field. Nothing public needs to change except for `.intersectPlane` now having a different signature than `.intersect`. It seems that it would behoove us to add `Plane.fromCartesian4` and probably `Plane.UNIT_X/Y/Z` or similar.

pjcozzi commented:
All sounds good.

kainino0x commented:
Ready for review. @bagnell @pjcozzi

kainino0x commented:
Made minor changes to `OrientedBoundingBox.fromRectangleTangentPlane`. Also changed check so that now bounding boxes are used as high as level 1.

mramato commented on CHANGES.md:
```diff
@@ -3,12 +3,16 @@ Change Log
 
+* Deprecated
+  * Deprecated `AxisAlignedBoundingBox.intersect` and `BoundingSphere.intersect`.  These will be removed in 1.13.  Use `.intersectPlane` and, if necessary, `Plane.fromCartesian4`.
+  * Deprecated the `ObjectOrientedBoundingBox` class.  It will be removed in 1.12.  If possible, `OrientedBoundingBox`, which is very similar.
```
Grammar.  I would just say "Use `OrientedBoundingBox` instead."

mramato commented on CHANGES.md:
```diff
@@ -3,12 +3,16 @@ Change Log
 * `Model` can now load Binary glTF from a Uint8Array.
 * Added a new camera mode for horizon views. When the camera is looking at the horizon and a point on terrain above the camera is picked, the camera moves in the plane containing the camera position, up and right vectors.
+* Added `Matrix2`/`Matrix3`/`Matrix4.ZERO` constants for zero matrices.
 
```
Since it's new, add a line like "Added `OrientedBoundingBox` which ..." explaining what it's for.

mramato replied:
Same for any other new functions or objects you added.

kainino0x replied:
Should I do that for `.intersectPlane` even though it's pretty much the same as `.intersect`?

mramato replied:
If they are basically the same, (and you already mentioned it in the deprecated section) then it's probably just noise for less-used functions like this.  If it were a popular part of the API, I would have said otherwise.

bagnell commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -186,6 +201,75 @@ define([
+
+        var ray = projectPointOntoPlaneRay;
+        ray.origin = cartesian;
+        Cartesian3.normalize(this._plane.normal, ray.direction);
```
The plane normal should already be normalized.

kainino0x commented:
I'm pushing `fixup` commits which will be rebased later. One of these also fixes the jsHint run.

mramato commented:
> I'm pushing "fixup" commits which will be rebased later. One of these also fixes the jsHint run.

While you can rebase if you want to, we don't really curate the commit history so it's not really a requirement.  I'm pretty sure GitHub keeps the old comments around even if you rebase.  If it doesn't, than that's probably a reason not to rebase.

kainino0x commented:
It does keep them; it just hides them so that it's easier to see what's probably been handled.

pjcozzi commented on CHANGES.md:
```diff
@@ -3,12 +3,21 @@ Change Log
 ### 1.11 - 2015-07-01
 
+* Deprecated
+  * Deprecated `AxisAlignedBoundingBox.intersect` and `BoundingSphere.intersect`.  These will be removed in 1.13.  Use `.intersectPlane` and, if necessary, `Plane.fromCartesian4`.
```
Please submit an issue for this and another for the OOBB class (for example, see #2724).

pjcozzi commented on CHANGES.md:
```diff
@@ -3,12 +3,21 @@ Change Log
 * Added a new camera mode for horizon views. When the camera is looking at the horizon and a point on terrain above the camera is picked, the camera moves in the plane containing the camera position, up and right vectors.
+* Added `Matrix2`/`Matrix3`/`Matrix4.ZERO` constants for zero matrices.
+* Added `Matrix2`/`Matrix3.multiplyByScale` for multiplying against non-uniform scales.
+* Added `EllipsoidTangentPlane.projectPoint(s)ToNearestOnPlane` to project 3D points to the nearest 2D point on an `EllipsoidTangentPlane`.
```
Say something like: "Added `projectPointToNearestOnPlane` and `projectPointsToNearestOnPlane` to `EllipsoidTangentPlane`.  The `(s)` is confusing.

pjcozzi commented on CHANGES.md:
```diff
@@ -3,12 +3,21 @@ Change Log
+* Added `Matrix2`/`Matrix3`/`Matrix4.ZERO` constants for zero matrices.
+* Added `Matrix2`/`Matrix3.multiplyByScale` for multiplying against non-uniform scales.
+* Added `EllipsoidTangentPlane.projectPoint(s)ToNearestOnPlane` to project 3D points to the nearest 2D point on an `EllipsoidTangentPlane`.
+* Added `OrientedBoundingBox` which is an optimized/simplified bounding box for optimization use.
```
Sounds awkward: "**optimized**/simplified bounding box for **optimization** use."

We could probably just say "Added `OrientedBoundingBox`" since 99% of users will not use this directly in the near future.

pjcozzi commented on CHANGES.md:
```diff
@@ -3,12 +3,21 @@ Change Log
+* Added `EllipsoidTangentPlane.projectPoint(s)ToNearestOnPlane` to project 3D points to the nearest 2D point on an `EllipsoidTangentPlane`.
+* Added `OrientedBoundingBox` which is an optimized/simplified bounding box for optimization use.
+* Added `EllipsoidTangentPlane.extentsToOrientedBoundingBox` to compute an `OrientedBoundingBox` from extents within an `EllipsoidTangentPlane`.
+* Use `OrientedBoundingBox` when rendering terrain for improved camera-view culling optimization.
```
This should be something like: "Improve terrain rendering performance by x% and loading performance by y% by adding culling optimizations."

The current wording is from an implementation perspective; users care about how things effect them.

pjcozzi commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -197,6 +198,25 @@ define([
+     *                      on the opposite side, and {@link Intersect.INTERSECTING} if the box
+     *                      intersects the plane.
+     */
+    AxisAlignedBoundingBox.intersect = function(box, plane) {
```
Use `deprecationWarning`.  See https://github.com/AnalyticalGraphicsInc/cesium/wiki/Deprecation-Guide.

Same comment throughout for anything else that is deprecated.

kainino0x replied:
Right, I forgot about that.

Since AABB.prototype.intersect calls AABB.intersect, should I only add it in AABB.intersect?

pjcozzi replied:
That is fine for this case.

pjcozzi commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -210,6 +230,20 @@ define([
+     *                      on the opposite side, and {@link Intersect.INTERSECTING} if the box
+     *                      intersects the plane.
+     */
+    AxisAlignedBoundingBox.prototype.intersectPlane = function(plane) {
```
On core math types, we usually only have prototype versions of functions for things like `clone` and `equals`.

Do we need this?  Is this because `intersectPlane` is being used polymorphically?  If so, this is fine.

kainino0x replied:
Yes, intersectPlane is being used polymorphically in the frustum plane tests.

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -417,11 +425,19 @@ define([
 
+        var rectangle = provider._tilingScheme.tileXYToRectangle(x, y, level);
+        var boundingOBB;
+        if (rectangle.width < CesiumMath.PI_OVER_TWO + 0.001) {
```
Use the epsilon values in `Math`.  You might want to also add a comment as to why we're using the epsilon as it wasn't clear to me at first glance (so we don't skip a level it looks like).

Also, shouldn't we check the `height` too?  It is probably a bad long-term move to assume `height <= width`.

kainino0x replied:
I left out the `height` check because I think it will handle any height (up to pi) decently well.

pjcozzi replied:
OK with me if we have a test for it.

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -417,11 +425,19 @@ define([
+        var boundingOBB;
+        if (rectangle.width < CesiumMath.PI_OVER_TWO + 0.001) {
+            var tangentPlane = new EllipsoidTangentPlane(center, provider._tilingScheme.ellipsoid);
+            boundingOBB = OrientedBoundingBox.fromRectangleTangentPlane(rectangle, tangentPlane, minimumHeight, maximumHeight);
```
This is awkward.  How is:
```javascript
boundingOBB = OrientedBoundingBox.fromRectangle(rectangle, minimumHeight, maximumHeight, provider._tilingScheme.ellipsoid);
```
Only `rectangle` would be required.  `EllipsoidTangentPlane` would be used in the OBB.  That dependency should be fine.

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -91,6 +95,17 @@ define([
+        },
+
+        /**
+         * Gets the plane which is tangent to the ellipsoid.
```
This should be `@readonly`.

Also mention it in CHANGES.md.

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -117,7 +132,7 @@ define([
 
     /**
-     * Computes the projection of the provided 3D position onto the 2D plane.
+     * Computes the projection of the provided 3D position onto the 2D plane, radially outward from the global origin.
```
Probably "WGS84 coordinate system origin" here and below.  WGS84 is our global coordinate system, but "global" is probably not explicit enough here.

kainino0x replied:
"... radially outward from the {@link EllipsoidTangentPlane.ellipsoid} coordinate system origin."?

pjcozzi replied:
OK.

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -186,6 +201,75 @@ define([
+        var length = cartesians.length;
+        for ( var i = 0; i < length; i++) {
+            var p = this.projectPointToNearestOnPlane(cartesians[i], result[count]);
+            if (defined(p)) {
```
The doc should mention that the array will not include points that could not be projected; otherwise, user may assume that the returned array has the same length as the input array.

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -186,6 +201,75 @@ define([
+            result.y = y;
+            return result;
+        }
+        return undefined;
```
Mention that this could happen in the reference doc.

kainino0x replied:
I copied this check from the `projectPointOntoPlane` - but I'm realizing that doesn't make sense since this projection can never mathematically fail. I'll just change that in both places.

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -186,6 +201,75 @@ define([
+        }
+        //>>includeEnd('debug');
+
+        var ray = projectPointOntoPlaneRay;
```
I know this was already here, but can you prefix this and the others with `scratch`?

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -230,5 +314,56 @@ define([
+    /**
+     * Computes an OrientedBoundingBox given extents in the east-north-up space of the tangent plane.
+     *
+     * @param {Number} minX Minimum x extent in tangent plane space.
```
With the exception of local identifiers (and some rare cases like WGS84), do not use abbreviations.

Spell out minimum and maximum here.

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -230,5 +314,56 @@ define([
+     */
+    EllipsoidTangentPlane.prototype.extentsToOrientedBoundingBox = function(minX, maxX, minY, maxY, minZ, maxZ, result) {
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(minX)) { throw new DeveloperError('minX is required.'); }
```
We don't bunch up `if` blocks like this (you could also just use the auto code formatter in Eclipse).

To make this less redundant, you could just have one `if` statement with `||`s.  This exception message will not be as precise, but that is fine since it is a rare case.

pjcozzi commented on Source/Core/EllipsoidTangentPlane.js:
```diff
@@ -230,5 +314,56 @@ define([
+     * @param {OrientedBoundingBox} [result] The object onto which to store the result.
+     * @returns {OrientedBoundingBox} The modified result parameter or a new OrientedBoundingBox instance if one was not provided.
+     */
+    EllipsoidTangentPlane.prototype.extentsToOrientedBoundingBox = function(minX, maxX, minY, maxY, minZ, maxZ, result) {
```
This function probably belongs on `OrientedBoundingBox` as a `fromXXX` function.

kainino0x replied:
I figured that would probably be the case, but it does require access to the private `_xAxis`/`_yAxis`. Should I instead add a way to get those? (a property

pjcozzi replied:
Add `@readonly` get properties using [defineProperties](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/defineProperties.js).

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,302 @@
+
+    /**
+     * Creates an instance of an OrientedBoundingBox.
+     * An OrientedBoundingBox model of an object or set of objects, is a closed volume (a cuboid), which completely contains the object or the set of objects.
```
Do you want to rewrite this old doc?  It is not useful for our users.  Just say it is convex and can provide a tighter fix than spheres and AABBs.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,302 @@
+     * @see OrientedBoundingBox.fromBoundingRectangle
+     * @see BoundingSphere
+     * @see BoundingRectangle
+     * @see ObjectOrientedBoundingBox
```
Drop this.  It is deprecated.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,302 @@
+     * @example
+     * // Create an OrientedBoundingBox using a transformation matrix, a position where the box will be translated, and a scale.
+     * var center = new Cesium.Cartesian3(1,0,0);
+     * var halfAxes = Cesium.Matrix3.clone(Cesium.Matrix3.fromScale(new Cartesian3(1.0, 3.0, 2.0)));
```
`Cesium.Matrix3.clone` is not needed here.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,302 @@
+     * // Compute an object oriented bounding box enclosing two points.
+     * var box = Cesium.OrientedBoundingBox.fromBoundingRectangle(boundingRectangle, 0.0);
+     */
+    OrientedBoundingBox.fromBoundingRectangle = function(boundingRectangle, rotation, result) {
```
Hmm, maybe mark this as `@private` until we know we need it.  Or just remove it since we have the other file.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,302 @@
+     *
+     * @param {Rectangle} rectangle The valid rectangle used to create a bounding box.
+     * @param {EllipsoidTangentPlane} tangentPlane A tangent plane with which the bounding box will be aligned.
+     * @param {Number} [minHeight=0.0] The minimum height (elevation) within the tile.
```
Spell out `min` and `max` here and throughout for public APIs.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,302 @@
+            throw new DeveloperError('tangentPlane is required');
+        }
+        
+        if (rectangle.width < 0.0 || rectangle.width > CesiumMath.PI) {
```
Document these `@exception`s.

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,302 @@
+                            Plane.getPointDistance(plane, cornersCartesianScratch[3]));
+        var maxZ = maxHeight;  // Since the tangent plane touches the surface at height = 0, this is okay
+
+        return tangentPlane.extentsToOrientedBoundingBox(minX, maxX, minY, maxY, minZ, maxZ, result);
```
As I commented above, why not roll this function into here?

pjcozzi commented on Source/Core/Plane.js:
```diff
@@ -95,6 +97,33 @@ define([
+     * @param {Plane} [result] The object onto which to store the result.
+     * @returns {Plane} A new plane instance or the modified result parameter.
+     */
+    Plane.fromCartesian4 = function(coefficients, result) {
```
Mention this in CHANGES.md (outside of the deprecated section).

pjcozzi commented on Source/Core/Plane.js:
```diff
@@ -119,5 +148,29 @@ define([
+     * @type {Plane}
+     * @constant
+     */
+    Plane.ORIGIN_XY_PLANE = freezeObject(new Plane(Cartesian3.UNIT_Z, 0.0));
```
Add these to CHANGES.md.

pjcozzi commented on Source/Core/QuantizedMeshTerrainData.js:
```diff
@@ -44,6 +46,7 @@ define([
      * @param {Number} options.minimumHeight The minimum terrain height within the tile, in meters above the ellipsoid.
      * @param {Number} options.maximumHeight The maximum terrain height within the tile, in meters above the ellipsoid.
      * @param {BoundingSphere} options.boundingSphere A sphere bounding all of the vertices in the mesh.
+     * @param {OrientedBoundingBox} [options.boundingOBB] An OrientedBoundingBox bounding all of the vertices in the mesh.
```
Add to CHANGES.md.

pjcozzi commented on Source/Core/TerrainMesh.js:
```diff
@@ -25,8 +25,9 @@ define([
       *                     scaled space, and used for horizon culling.  If this point is below the horizon,
       *                     the tile is considered to be entirely below the horizon.
       * @param {Number} [vertexStride=6] The number of components in each vertex.
+      * @param {OrientedBoundingBox} [boundingOBB=undefined] A bounding sphere that completely contains the tile.
```
Drop `=undefined`, it is implicit.

pjcozzi commented on Source/Core/TerrainMesh.js:
```diff
@@ -82,6 +83,12 @@ define([
+         * A bounding box that completely contains the tile.
+         * @type {OrientedBoundingBox}
+         */
+        this.boundingOBB = boundingOBB;
```
Spell it out and add it to CHANGES.md.

pjcozzi commented on Source/Renderer/DrawCommand.js:
```diff
@@ -32,6 +32,17 @@ define([
+         *
+         * @see DrawCommand#debugShowBoundingVolume
+         */
+        this.boundingOBB = options.boundingOBB;
```
Why can't we just use `boundingVolume`?  Because we transform these to WGS84 in Scene.js and we won't know what scratch type to use?

kainino0x replied:
I plan to change it over to use `boundingVolume`. I had a discussion about this with @bagnell yesterday and it seems to make more sense.

pjcozzi replied:
OK.

kainino0x replied:
In the debug code which draws the bounding volumes for the inspector, there is different code for spheres and boxes. Is there a way to check there which it is if they're combined? That would be bad design in general, but seems okay for debug code (?)

pjcozzi replied:
Yes, it is fine for debug code.  Just check to see if one of the unique properties is defined and include a comment saying why it is checking.

kainino0x replied:
Since these are attached to the DrawCommand, there are other places in the code where it doesn't work, e.g. in `Scene.js` `createPotentiallyVisibleSet`:

```
BoundingSphere.computePlaneDistances(boundingVolume, position, direction, distances);
```

I could implement `computePlaneDistances` for `OrientedBoundingBox`, but I'm not clear on whether we really want `createPotentiallyVisibleSet` to be calling that.

pjcozzi replied:
@kainino0x and I discussed offline, we are implementing `computePlaneDistances`.

pjcozzi commented on Source/Scene/GlobeSurfaceTile.js:
```diff
@@ -8,6 +8,7 @@ define([
         '../Core/defined',
         '../Core/defineProperties',
         '../Core/IntersectionTests',
+        '../Core/OrientedBoundingBox',
```
This include isn't needed since the type is not explicitly used in this file.

pjcozzi commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -395,7 +399,7 @@ define([
-        var boundingVolume = surfaceTile.boundingSphere3D;
+        var boundingVolume = defaultValue(surfaceTile.boundingOBB, surfaceTile.boundingSphere3D);
 
         if (frameState.mode !== SceneMode.SCENE3D) {
```
This reminds me - have you tested in 2D and Columbus view?  Make sure to visually inspect too.

kainino0x replied:
I haven't done anything to make it use OBBs in 2D and Columbus at all - it still falls back to using bounding spheres in exactly the same way as before. I guess that's something to add in.

pjcozzi replied:
Yes, but we can do it later.  Let's test the performance improvement first and then see how involved this is.  It could be easy, but I am not familiar with all the code.

pjcozzi commented on Specs/Core/OrientedBoundingBoxSpec.js:
```diff
@@ -0,0 +1,442 @@
+        var rotScale = new Matrix3(0.0, 0.0, 0.5 * (1.0 - Math.SQRT1_2), 0.0, 0.0, 0.0, 0.0, Math.SQRT1_2, 0.0);
+        expect(box.halfAxes).toEqualEpsilon(rotScale, CesiumMath.EPSILON15);
+    });
+    /**
```
You don't need to include these comments in tests.

pjcozzi commented:
@kring do you want to look at the changes in the terrain files?

pjcozzi commented:
@kainino0x that is all I have for this round.  Let me know when this is ready and I'll do another review, run the tests, etc.

kring commented:
This looks straightforward to me from a terrain perspective.

One question, though.  Are you sure that `OrientedBoundingBox.fromRectangleTangentPlane` is guaranteed to actually bound the rectangle?  It looks like it's projecting just the corner points of the extent to a tangent plane and using that to form a box.  But the extent itself can easily extend outside the box defined by the corners.  For an extreme example, imagine an extent that goes from the North pole to the South pole.  The bounding box will need to be pretty wide to contain the extent at the equator, but the four corners projected onto a tangent plane will only form a narrow box.  Especially since you appear to be projecting along the plane normal not the ellipsoid surface normal.

I probably just missed where this is accounted for, but I wanted to mention it just in case.

kainino0x commented:
@kring You're right, I never tested for a tile which spans the equator, which is the case where that becomes a problem. I was going to add a test for it, but I hadn't gotten to it yet - now that I have, I'll find a good way to solve this.

kainino0x commented:
Ready for round 2 of reviews. Currently: working on a better place to put debug rendering.

pjcozzi commented:
This is without `DeveloperError` optimized out (see https://github.com/AnalyticalGraphicsInc/cesium/wiki/Contributor's-Guide), but we should optimize (remove) our use of `getColumn` since it is a hot spot when rendering in the steady state:

![image](https://cloud.githubusercontent.com/assets/782098/8140016/27eb9a66-1126-11e5-98d5-d5a4753b8962.png)

pjcozzi commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,332 @@
+     * @exception {DeveloperError} rectangle.width must be between 0 and pi.
+     * @exception {DeveloperError} rectangle.height must be between 0 and pi.
+     */
+    OrientedBoundingBox.fromEllipsoidRectangle = function(ellipsoid, rectangle, minimumHeight, maximumHeight, result) {
```
`OrientedBoundingBox.fromRectangle` would be a fine name.

Make the order
```
rectangle, minimumHeight, maximumHeight, ellipsoid, result
```
As I suggested previously, default `ellipsoid` to WGS84 like we do everywhere else.

pjcozzi commented on Specs/Core/OrientedBoundingBoxSpec.js:
```diff
@@ -0,0 +1,487 @@
+    });
+
+    /*
+    it('extentsToOrientedBoundingBox works with a result parameter', function() {
```
We don't include commented out code in code we merge to master.

kring commented:
>  I never tested for a tile which spans the equator, which is the case where that becomes a problem.

I just used that as an example because it's easy to visualize.  I think there are other cases where the corner points are insufficient.  For example, consider an extent that spans a wide longitude range.  All the latitudes _on the ellipsoid_ at the North and South edge lie in a plane, but this is _not_ true of points displaced from the surface along the ellipsoid surface normal (as terrain heights are).  And even if these points did still lie in a plane (i.e. because they're on the ellipsoid surface), this plane would not be perpendicular to the tangent plane.  This means that points of equal height along the Northern and Southern boundary, when projected onto the tangent plane, do not have the same Y coordinate in the plane.  In the Northern hemisphere, on the Southern boundary of the extent, the middle point extends farther from the plane center than the corners do.  The opposite is true in the Southern hemisphere.

You can see this clearly in the "bulge" at the bottom of this extent:
![image](https://cloud.githubusercontent.com/assets/924374/8144046/36570f30-120d-11e5-9c29-12ff390a8aa7.png)

So I _think_ it's sufficient to define your OBB using the four corners, the West and East longitudes at the equator if the extent crosses the equator, and the middle points at the Northern and Southern boundary. (unless I just lack imagination in thinking of other cases like the two I've described now!)  This is all assuming a spheroid.  On a general ellipsoid I have no idea!

kainino0x commented:
I believe that what you described is exactly what I am doing now. On an ellipsoid of revolution (a = b != c) I *think* this is sufficient, though for a tri-axial (a != b != c) ellipsoid the north and south center points probably aren't sufficient.

kring commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,332 @@
+        var lonCenter = tangentPointCartographic.longitude;
+        var latCenter = tangentPointCartographic.latitude;
+        perimeterSW.latitude = perimeterSC.latitude = perimeterSE.latitude = rectangle.south;
+        perimeterCW.latitude = perimeterCE.latitude = latCenter;
```
I think you want to use the equator here if the extent crosses it, not the center latitude.  If the extent doesn't cross the equator you probably don't need `perimeterSW` and `perimeterCW` at all.

kainino0x replied:
I should clarify here that "C" refers to whatever point taken as the tangent point - which will be at the equator if the rectangle spans the equator.

kring replied:
Ok, sorry, I had missed that you're using the equator to define the tangent plane when the rectangle crosses it.  Though, I don't understand why you're doing that.  Wouldn't you get a "better" OBB if you used the center of the rectangle instead of the equator for a rectangle that goes from, say, -4 degrees latitude to 80 degrees latitude?

And speaking of center, averaging the latitudes and longitudes of the sides doesn't give you a very good center for the rectangle (it's skewed too heavily toward the poles).  You might consider using the midpoint latitude code I just added to `rectangleCameraPosition3D` in `Camera`.  It finds the midpoint along the geodesic, which I think would be better here, too.

kainino0x replied:
That's a good question. I did it that way because I wanted to avoid any questions about whether the oblateness of the ellipsoid would make the equator-aligned points not-quite-right. But I am pretty sure that the center of the rectangle will still be valid. (Again - at least on an ellipsoid of revolution. I should put a check in for that.)

I agree that the latitudinal/longitudinal centers would not be very accurate. For this reason, I'm actually not using that as the center of the bounding box - it gets shifted over to the mean of the min and max in each local axis. However, using the midpoint along the geodesic would likely result in a better bounding box alignment.

kring commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -0,0 +1,332 @@
+        var minZ = Math.min(Plane.getPointDistance(plane, perimeterCartesianScratch[0]),
+                            Plane.getPointDistance(plane, perimeterCartesianScratch[1]),
+                            Plane.getPointDistance(plane, perimeterCartesianScratch[2]),
+                            Plane.getPointDistance(plane, perimeterCartesianScratch[3]));
```
Only 0-3?  I'm pretty sure 4-7 can affect the minZ, too.

kainino0x replied:
This is a change that I meant to amend on but didn't add before stopping work on Friday. It's meant to be 0, 2, 4, 6 (the corner points at minimum height).

kring commented:
> I believe that what you described is exactly what I am doing now.

No quite.  I added some comments to the diff.

kainino0x commented:
I believe I've addressed @kring's concerns, now (except for some unnecessary extent calculations). I've also done some more testing and fixed a few blatant bugs which weren't caught before.

There are still no hard performance results as I haven't figured out how to make it request a deterministic(-ish) number of terrain tiles from the server. However, it seems that in a case like [this high-altitude low-angle view of Mount Pinatubo](http://localhost:8080/Build/Apps/CesiumViewer/index.html?stats=true&view=119.7954832286864%2C14.625866129373039%2C32421.396898127718%2C46.017725109639954%2C-21.092747155669294%2C0.07233855930825417) the number of terrain tiles requested from the server is down something like 5-15% from before.


Ready for another round of reviewing. @pjcozzi, @bagnell?

kring commented:
> I haven't figured out how to make it request a deterministic(-ish) number of terrain tiles from the server

Rather than looking at the number of requests, just wait until you hit the steady state (everything done loading) and look at the total number of tiles rendered.  Set `scene.globe._surface._debug.enableDebugOutput = true` and it will print the number of tiles rendered to the console.

pjcozzi commented:
@kainino0x merge in master.

pjcozzi commented:
@kainino0x given everything else we have coming up, I suggest we add these to the terrain roadmap and handle them later in a separate PR:
> Future:
> [ ] Hook up to heightmapped terrain (and flat terrain, if that's separate); not just quantized meshes
> [ ] Hook up to 2D and Columbus view rendering modes

pjcozzi commented:
How do I visualize the bounding box?  I still get bounding spheres in the inspector:

![image](https://cloud.githubusercontent.com/assets/782098/8206268/e6e0cc0a-14c5-11e5-8a66-f6c7f75ddd84.png)

I'm pretty sure I am running your branch since I have the new `OrientedBoundingBox` type.

pjcozzi commented on CHANGES.md:
```diff
@@ -10,6 +13,17 @@ Change Log
+* Added `EllipsoidTangentPlane.xAxis`/`yAxis`/`zAxis` properties to get the local coordinate system of the tangent plane.
+* Add `QuantizedMeshTerrainData` constructor argument `orientedBoundingBox`.
+* Add `TerrainMesh.orientedBoundingBox` which holds the `OrientedBoundingBox` for the mesh for a single terrain tile.
+* Use `OrientedBoundingBox` when rendering terrain to improve performance of terrain rendering (by **TODO**%) and loading (by **TODO**%).
```
@kainino0x please remember to update this TODO once you have the performance numbers.

pjcozzi replied:
@kainino0x can you update this now that you ran the tests?

pjcozzi commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -907,7 +962,13 @@ define([
-            command.debugShowBoundingVolume = (tile === tileProvider._debug.boundingSphereTile);
+            if (tile === tileProvider._debug.boundingSphereTile) {
+                if (defined(surfaceTile.orientedBoundingBox)) {
+                    createDebugOrientedBoundingBox(surfaceTile.orientedBoundingBox, Color.RED).update(context, frameState, commandList);
```
Here and below leak the debug primitive on each frame.  We should keep a reference to it so we can call `destroy`.  Stop by and I can explain more.

pjcozzi replied:
I don't think we addressed this.

kainino0x replied:
Take a look at the large block of added code above - shouldn't that handle it? It no longer always creates everything - it only does so if the selected tile changes.

pjcozzi replied:
There is only one reference to the debug primitive.  Don't we need a reference per tile?

kainino0x replied:
Since (at least currently) there is only one debug primitive rendered at once, this seemed easier since I wouldn't have to muck around inside some tile class. This should avoid recreating it at least between frames while one tile is selected.

pjcozzi replied:
Ah, you are right.  Can you add a comment to the code that describes this?

pjcozzi commented:
I believe the only outstanding issues are:

* Debug code
* Performance numbers

Once we merge this, we should write a short tech blog post that includes:
* sphere vs box screenshots (maybe state the area as well)
* before/after suspended LOD screenshots
* before/after performance numbers for certain views
* Discussion of how this becomes less important deeper in the tree (assuming that is true)
* Discussion on the CPU overhead impact - if we can tell in the profiler - no need to cover the spatial coherence optimization because we are not using it yet
* No need to go into detail on how to compute the box

kainino0x commented:
@pjcozzi
> How do I visualize the bounding box? I still get bounding spheres in the inspector:

Do you have terrain enabled? It still doesn't use bounding boxes if terrain is off.

pjcozzi commented:
Ah, you are right!

kainino0x commented:
@kring
> Rather than looking at the number of requests, just wait until you hit the steady state (everything done loading) and look at the total number of tiles rendered. Set `scene.globe._surface._debug.enableDebugOutput = true` and it will print the number of tiles rendered to the console.

Thanks for the tip - this is useful for investigating the rendering performance. In terms of network performance, though, is there any way I can use this to measure a reduction in loading time?

pjcozzi commented:
Zoomed in to "Seneca Rocks, WV" at level 20, I am getting a bounding sphere:

![image](https://cloud.githubusercontent.com/assets/782098/8215451/d930f944-14fd-11e5-8cf6-ed31dc90006f.png)

But, for example, I get the box at level 5:

![image](https://cloud.githubusercontent.com/assets/782098/8215463/f45ed0e2-14fd-11e5-879b-52ead0db1463.png)

pjcozzi commented:
Also, I suspect this is because the tiles are being up-sampled from the deepest level available so we are missing a place where we should compute the OBB.

kainino0x commented:
I was seeing a problem like this - but I thought it had gone away when I did some reworking of the terrain integration changes. This is probably the same. That does seem like a reasonable suspicion, I just haven't figured out what that place might be, because I haven't figured out the rule for when it happens.

pjcozzi commented:
It happens when zooming in close and no more real terrain is available.  Search for upsample or downsample.  @kring will know where of the top of his head or I can trace through it if you have trouble finding it.

kainino0x commented:
This simple fix seems to work. I will have to do a bit more testing to see if I can still find similar issues. I did at one point have very similar changes in this file, but I probably didn't implement them correctly (I removed them because they weren't fixing this problem). This change was very small, though, since I had already wired it up to work last time I tried.

kainino0x commented:
There have been a few changes that should be reviewed but other than these I don't know of anything I might change.

I added an adjustment to expand the bounding box to cover the tile skirt. This seems like something we should have, although it seems to sometimes pretty significantly increase the number of tiles rendered (279 to 298 compared to 330 for the sphere in one case). Any comments?

kainino0x commented:
I'm working on a blog post draft here: https://github.com/kainino0x/draft-obb-blog-post

kring commented:
> I added an adjustment to expand the bounding box to cover the tile skirt. 

It shouldn't be necessary to account for the skirts, since they're only intended to fill the vertical space between tiles.  It's hard to imagine a situation where only the skirt is visible but the rest of the tile is not.  You haven't seen such a case have you?

pjcozzi commented:
@kainino0x and I talked offline, we are not going to include the skirts, but we'll comment it in the code incase we ever see cracks.

pjcozzi commented:
> I'm working on a blog post draft here: https://github.com/kainino0x/draft-obb-blog-post

Looks great.  I will do some minor edits and publish next week.  Is there anything else you wanted to add/change?

kainino0x commented:
I'll change the numbers/screenshots a bit to reflect the removal of the skirt change. I'll push that to the repo, so you'll have it when you go to do the editing.

kainino0x commented:
For the record, the case in which the skirts hurt the performance was this one: ["Toward horizon, low altitude"](http://cesiumjs.org/Cesium/Build/Apps/CesiumViewer/index.html?view=120.28867831296651%2C15.087807528789323%2C4406.192801986939%2C46.14351181495515%2C-20.429229458524063%2C0.0742423709965139). Went from 279 to 298 tiles rendered.

kainino0x commented:
In the blog post, should I mention why the bounding box appears not to contain the tile, because of the skirt sticking out?

pjcozzi commented:
Yes, I think that is an interesting point to make: the box doesn't full bound the geometry, but, in practice, we don't expect it to generate artifacts since the view position/direction would never be such that it would matter.

pjcozzi commented:
@kainino0x let me know when the blog post is ready and I will copyedit and publish this week.

Also, let me know this is ready to merge.

pjcozzi commented:
@kainino0x also, remind me why we decided not to use the OBB for binning into multiple frustums?  I want to keep some notes in case we decide to do it later.

kainino0x commented:
@pjcozzi I just pushed a commit to the blog post which I had forgotten to push on Friday. Blog and PR should be ready.

I don't know that we made a specific decision not to use the OBB for binning into multiple frustums. It's possible that we just decided when it came up that I should keep working on other parts first.

kainino0x commented:
Oh, I also forgot to push the heightmapped terrain commit to this repository on Friday. Please review that short commit. @pjcozzi

pjcozzi commented:
@kainino0x can you merge in master?

This looks good.  @kainino0x are you OK with us merging this so it can sit in master for a few days before the 1.11 release on July 1?  Any additional OBB optimizations can come in a separate PR.

kainino0x commented:
Sure. I'll sync with master when I get home today.

kainino0x commented:
Done. I'm seeing a bunch of errors but they are also there on master. Plus the timeout ones that I see regularly.

mramato commented:
The failures in master are a website issue.  I'm putting in a fix now.

pjcozzi commented:
Thanks @kainino0x.  If we want to tweak this before the 1.11 release, we have until early next week.
